pkg/services/sqlstore/migrations/ualert/ualert.go

```diff
@@ -876,3 +879,104 @@ func (c updateRulesOrderInGroup) Exec(sess *xorm.Session, migrator *migrator.Mig
 	}
 	return nil
 }
+
+func AddExternalAlertmanagerToDatasourceMigration(mg *migrator.Migrator) {
+	mg.AddMigration("migrate external alertmanagers to datsourcse", &externalAlertmanagerToDatasources{})
+}
+
+type externalAlertmanagerToDatasources struct {
+	migrator.MigrationBase
+}
+
+// Just copy old struct code to use Alertmanagers as new one is not getting it.
+type AdminConfiguration struct {
+	ID    int64 `xorm:"pk autoincr 'id'"`
+	OrgID int64 `xorm:"org_id"`
+
+	Alertmanagers []string
+
+	// SendAlertsTo indicates which set of alertmanagers will handle the alert.
+	SendAlertsTo ngmodels.AlertmanagersChoice `xorm:"send_alerts_to"`
+
+	CreatedAt int64 `xorm:"created"`
+	UpdatedAt int64 `xorm:"updated"`
+}
+
+func (e externalAlertmanagerToDatasources) SQL(dialect migrator.Dialect) string {
+	return "migrate external alertmanagers to datasource"
+}
+
+func (e externalAlertmanagerToDatasources) Exec(sess *xorm.Session, mg *migrator.Migrator) error {
+	var results []AdminConfiguration
+	err := sess.SQL("SELECT org_id, alertmanagers FROM ngalert_configuration").Find(&results)
+	if err != nil {
+		return err
+	}
+
+	for _, result := range results {
+		for _, am := range result.Alertmanagers {
```

Comment:
Should we do some kind of deduplication on the URL? If a system has two configured alertmanagers with exactly the same URL/credentials, should we collapse them into one datasource?

I ask because we've had duplication bugs with alertmanagers in the past. There are definitely still instances floating around with many duplicated URLs in this list.

Reply:
I agree, and think we should deduplicate on the URL. I just checked and neither the UI or API deduplicated external Alertmanagers.

Reply:
Yes! Your point is correct